### PR TITLE
br: acceleration of snapshot restore speed

### PR DESCRIPTION
**Pull request by benmaoer**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v8.0 (TiDB 8.0 versions)
- [ ] v7.6 (TiDB 7.6 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.4 (TiDB 7.4 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: br/br-snapshot-guide.md [hunk 1]
```diff
@@ -62,6 +62,11 @@ tiup br validate decode --field="end-version" \
 
 ## 恢复快照备份数据
 
+> **注意：**
+> - 在 BR v7.5.0 及之前版本中，每个 TiKV 节点的快照恢复速度约为 100 MiB/s
```
Comment (qiancai) [inline br/br-snapshot-guide.md:66]:
```suggestion
> **注意：**
>
> - 在 BR v7.5.0 及之前版本中，每个 TiKV 节点的快照恢复速度约为 100 MiB/s。
```

File: br/br-snapshot-guide.md [hunk 2]
```diff
@@ -62,6 +62,11 @@ tiup br validate decode --field="end-version" \
...
+> - 从 BR v7.6.0 开始，为了解决大规模 Region 场景下可能出现的恢复瓶颈问题，BR 支持以粗细粒度结合的方式打散 Region 进行恢复（实验特性）。可以通过指定命令行参数 `--granularity="coarse-grained"` 来启用此功能。
+> - 从 BR v8.0.0 版本开始，粗细粒度结合的方式打散 Region 进行数据恢复功能正式 GA，并默认启用。通过改进算法、批量创建库表、降低操作之间的相互影响等措施，快照恢复的速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2GiB/s ，能够在 1 小时内完成对 100TiB 数据的恢复。
```
Comment (qiancai) [inline br/br-snapshot-guide.md:68]:
```suggestion
> - 从 BR v8.0.0 版本开始，粗细粒度结合的方式打散 Region 进行数据恢复功能正式 GA，并默认启用。通过改进算法、批量创建库表、降低操作之间的相互影响等措施，快照恢复的速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2GiB/s，能够在 1 小时内完成对 100TiB 数据的恢复。
```

**Review by BornChanger — APPROVED** [review]

File: br/br-snapshot-guide.md [hunk 3]
```diff
@@ -62,6 +62,12 @@ tiup br validate decode --field="end-version" \
 
 ## 恢复快照备份数据
 
+> **注意：**
+>
+> - 在 BR v7.5.0 及之前版本中，每个 TiKV 节点的快照恢复速度约为 100 MiB/s。
+> - 从 BR v7.6.0 开始，为了解决大规模 Region 场景下可能出现的恢复瓶颈问题，BR 支持以粗细粒度结合的方式打散 Region 进行恢复（实验特性）。可以通过指定命令行参数 `--granularity="coarse-grained"` 来启用此功能。
+> - 从 BR v8.0.0 版本开始，粗细粒度结合的方式打散 Region 进行数据恢复功能正式 GA，并默认启用。通过改进算法、批量创建库表、降低操作之间的相互影响等措施，快照恢复的 sst 下载速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2GiB/s，端到端的恢复速度提升到3倍,  能够在 1 小时内完成对 100TiB 数据的恢复。
```
Comment (qiancai) [inline br/br-snapshot-guide.md:69]:
```suggestion
> - 从 BR v8.0.0 版本开始，通过粗粒度打散 Region 算法进行数据恢复的功能正式 GA，并默认启用。通过改进打散算法、批量创建库表、降低操作之间的相互影响等措施，快照恢复的 SST 下载速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2 GiB/s，端到端的恢复速度提升约 2.5 到 3 倍,  能够在 1 小时内完成对 100TiB 数据的恢复。
```

File: br/br-snapshot-guide.md [hunk 4]
```diff
@@ -205,7 +205,7 @@ TiDB 备份功能对集群性能（事务延迟和 QPS）有一定的影响，
 
 - TiDB 恢复的时候会尽可能打满 TiKV CPU、磁盘 IO、网络带宽等资源，所以推荐在空的集群上执行备份数据的恢复，避免对正在运行的业务产生影响。
 - 备份数据的恢复速度与集群配置、部署、运行的业务都有比较大的关系。在内部多场景仿真测试中，单 TiKV 存储节点上备份数据恢复速度能够达到 100 MiB/s。在不同用户场景下，快照恢复的性能和影响应以实际测试结论为准。
-- 从 v7.6.0 开始，BR 提供了一个实验性特性，允许通过指定命令行参数 `--granularity="coarse-grained"` 启用粗粒度的 Region 打散算法，加快大规模 Region 场景下的 Region 恢复速度。在这个方式下每个 TiKV 节点会得到均匀稳定的下载任务，从而充分利用每个 TiKV 节点的所有资源实现并行快速恢复。在实际案例中，大规模 Region 场景下，集群快照恢复速度最高提升约 10 倍。使用示例如下：
+- 从 v8.0.0 开始，BR 提供了粗粒度的 Region 打散特性，通过命令行参数 `--granularity="coarse-grained"` 控制，并默认启用，用于提升大规模 Region 场景下的 Region 恢复速度。在这个方式下每个 TiKV 节点会得到均匀稳定的下载任务，从而充分利用每个 TiKV 节点的所有资源实现并行快速恢复。同时，新增 `--tikv-max-restore-concurrency` 参数，用于控制 BR 恢复过程中针对每个 TiKV 的最大 download/ingest 可执行任务数量, 并通过控制作业队列的最大长度, 可以控制 BR 节点的内存消耗（该参数默认情况下，会根据集群配置自动调整，一般不需要特殊设置）。在实际案例中，大规模 Region 场景下，集群快照恢复速度最高提升约 3 倍。使用示例如下：
```
Comment (qiancai) [inline br/br-snapshot-guide.md:208]:
```suggestion
- BR 提供了粗粒度的 Region 打散算法，用于提升大规模 Region 场景下的 Region 恢复速度。该算法通过命令行参数 `--granularity="coarse-grained"` 控制，并默认启用。在这个方式下每个 TiKV 节点会得到均匀稳定的下载任务，从而充分利用每个 TiKV 节点的所有资源实现并行快速恢复。在实际案例中，大规模 Region 场景下，集群快照恢复速度最高提升约 3 倍。使用示例如下：
```

File: br/br-snapshot-guide.md [hunk 5]
```diff
@@ -62,6 +62,12 @@ tiup br validate decode --field="end-version" \
 
 ## 恢复快照备份数据
 
+> **注意：**
+>
+> - 在 BR v7.5.0 及之前版本中，每个 TiKV 节点的快照恢复速度约为 100 MiB/s。
+> - 从 BR v7.6.0 开始，为了解决大规模 Region 场景下可能出现的恢复瓶颈问题，BR 支持以粗细粒度结合的方式打散 Region 进行恢复（实验特性）。可以通过指定命令行参数 `--granularity="coarse-grained"` 来启用此功能。
+> - 从 BR v8.0.0 版本开始，通过粗粒度打散 Region 算法进行数据恢复的功能正式 GA，并默认启用。通过采用粗粒度打散 Region 算法、批量创建库表、降低 SST 文件下载和 Ingest 操作之间的相互影响、加速表统计信息恢复等改进措施，快照恢复的速度有大幅提升。在实际案例中，快照恢复的 SST 文件下载速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2 GiB/s，端到端的恢复速度大约提升 1.5 到 3 倍，并且能够在 1 小时内完成对 100 TiB 数据的恢复。
```
Comment (qiancai) [inline br/br-snapshot-guide.md:69]:
```suggestion
> - 从 BR v7.6.0 开始，为了解决大规模 Region 场景下可能出现的恢复瓶颈问题，BR 支持通过粗粒度打散 Region 的算法加速恢复（实验特性）。可以通过指定命令行参数 `--granularity="coarse-grained"` 来启用此功能。
> - 从 BR v8.0.0 版本开始，通过粗粒度打散 Region 算法进行数据恢复的功能正式 GA，并默认启用。通过采用粗粒度打散 Region 算法、批量创建库表、降低 SST 文件下载和 Ingest 操作之间的相互影响、加速表统计信息恢复等改进措施，快照恢复的速度有大幅提升。在实际案例中，快照恢复的 SST 文件下载速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2 GiB/s，端到端的恢复速度大约提升 1.5 到 3 倍，并且能够在 1 小时内完成对 100 TiB 数据的恢复。
```

File: br/br-snapshot-guide.md [hunk 6]
```diff
@@ -217,6 +217,13 @@ TiDB 备份功能对集群性能（事务延迟和 QPS）有一定的影响，
     --log-file restorefull.log
     ```
 
+- 从 v8.0.0 起，`br` 命令行工具新增 `--tikv-max-restore-concurrency` 参数，用于控制每个 TiKV 节点的最大 download 和 ingest 文件数量，并通过控制作业队列的最大长度，进而控制 BR 节点的内存消耗。每个 TiKV 节点的作业队列的最大长度与 `--tikv-max-restore-concurrency` 参数值成正比，比例系数为 32。
```
Comment (Leavrth) [inline br/br-snapshot-guide.md:220]: 与`--tikv-max-restore-concurrency` 参数值 和 tikv 数成正比

**Review by Leavrth — APPROVED** [review]

File: br/br-snapshot-guide.md (hunk 6 above)
Comment (qiancai) [inline br/br-snapshot-guide.md:220]:
```suggestion
- 从 v8.0.0 起，`br` 命令行工具新增 `--tikv-max-restore-concurrency` 参数，用于控制每个 TiKV 节点的最大 download 和 ingest 文件数量，并通过控制作业队列的最大长度，进而控制 BR 节点的内存消耗。作业队列的最大长度与 TiKV 数量成正比，与 `--tikv-max-restore-concurrency` 参数值成正比，比例系数为 32。
```

File: br/br-snapshot-guide.md [hunk 7]
```diff
@@ -62,6 +62,12 @@ tiup br validate decode --field="end-version" \
 
 ## 恢复快照备份数据
 
+> **注意：**
+>
+> - 在 BR v7.5.0 及之前版本中，每个 TiKV 节点的快照恢复速度约为 100 MiB/s。
+> - 从 BR v7.6.0 开始，为了解决大规模 Region 场景下可能出现的恢复瓶颈问题，BR 支持通过粗粒度打散 Region 的算法加速恢复（实验特性）。可以通过指定命令行参数 `--granularity="coarse-grained"` 来启用此功能。
+> - 从 BR v8.0.0 版本开始，通过粗粒度打散 Region 算法进行数据恢复的功能正式 GA，并默认启用。通过采用粗粒度打散 Region 算法、批量创建库表、降低 SST 文件下载和 Ingest 操作之间的相互影响、加速表统计信息恢复等改进措施，快照恢复的速度有大幅提升。在实际案例中，快照恢复的 SST 文件下载速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2 GiB/s，端到端的恢复速度大约提升 1.5 到 3 倍，并且能够在 1 小时内完成对 100 TiB 数据的恢复。
```
Comment (qiancai) [inline br/br-snapshot-guide.md:69]:
```suggestion
> - 从 BR v8.0.0 版本开始，通过粗粒度打散 Region 算法进行快照恢复的功能正式 GA，并默认启用。通过采用粗粒度打散 Region 算法、批量创建库表、降低 SST 文件下载和 Ingest 操作之间的相互影响、加速表统计信息恢复等改进措施，快照恢复的速度有大幅提升。在实际案例中，快照恢复的 SST 文件下载速度最高提升约 10 倍，单个 TiKV 节点的数据恢复速度稳定在 1.2 GiB/s，端到端的恢复速度大约提升 1.5 到 3 倍，并且能够在 1 小时内完成对 100 TiB 数据的恢复。
```

**Review by lilin90 — APPROVED** [review]
/lgtm

File: br/br-snapshot-guide.md [hunk 8]
```diff
@@ -217,6 +217,13 @@ TiDB 备份功能对集群性能（事务延迟和 QPS）有一定的影响，
     --log-file restorefull.log
     ```
 
+- 从 v8.0.0 起，`br` 命令行工具新增 `--tikv-max-restore-concurrency` 参数，用于控制每个 TiKV 节点的最大 download 和 ingest 文件数量。此外，通过调整此参数，可以控制作业队列的最大长度（作业队列的最大长度 = 32 * TiKV 节点数量 * `--tikv-max-restore-concurrency`），进而控制 BR 节点的内存消耗。
+
+    通常情况下，`--tikv-max-restore-concurrency` 会根据集群配置自动调整，无需手动设置。如果通过 Grafana 中的 **TiKV-Details** > **Backup & Import** > **Import RPC count** 监控指标发现 download 文件数量长时间接近于 0，而 ingest 文件数量一直处于上限时，说明 ingest 文件任务存在堆积，并且作业队列已达到最大长度。此时，可以采取以下措施来缓解积压问题：
+
+    - 设置 `--ratelimit` 参数来限制下载速度，以确保 ingest 文件任务有足够的资源。例如，当任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽大于 `x/2 MiB/s`，可以设置参数 `--ratelimit x/2`。如果任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽小于或等于 `x/2 MiB/s`，可以不设置参数 `--ratelimit`。
+    - 调高 `--tikv-max-restore-concurrency` 来增加作业队列的最大长度，缓解堆积问题。
```
Comment (qiancai) [inline br/br-snapshot-guide.md:225]:
```suggestion
    通常情况下，`--tikv-max-restore-concurrency` 会根据集群配置自动调整，无需手动设置。如果通过 Grafana 中的 **TiKV-Details** > **Backup & Import** > **Import RPC count** 监控指标发现 download 文件数量长时间接近于 0，而 ingest 文件数量一直处于上限时，说明 ingest 文件任务存在堆积，并且作业队列已达到最大长度。此时，可以采取以下措施来缓解积压问题：

    - 设置 `--ratelimit` 参数来限制下载速度，以确保 ingest 文件任务有足够的资源。例如，当任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽大于 `x/2 MiB/s`，可以设置参数 `--ratelimit x/2`。如果任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽小于或等于 `x/2 MiB/s`，可以不设置参数 `--ratelimit`。
    - 调高 `--tikv-max-restore-concurrency` 来增加作业队列的最大长度，缓解堆积问题。
```
```suggestion
    通常情况下，`--tikv-max-restore-concurrency` 会根据集群配置自动调整，无需手动设置。如果通过 Grafana 中的 **TiKV-Details** > **Backup & Import** > **Import RPC count** 监控指标发现 download 文件数量长时间接近于 0，而 ingest 文件数量一直处于上限时，说明 ingest 文件任务存在堆积，并且作业队列已达到最大长度。此时，可以采取以下措施来缓解积压问题：

    - 设置 `--ratelimit` 参数来限制下载速度，以确保 ingest 文件任务有足够的资源。例如，当任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽大于 `x/2 MiB/s`，可以设置参数 `--ratelimit x/2`。如果任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽小于或等于 `x/2 MiB/s`，可以不设置参数 `--ratelimit`。
    - 调高 `--tikv-max-restore-concurrency` 来增加作业队列的最大长度，缓解堆积问题。
```
```suggestion
    通常情况下，`--tikv-max-restore-concurrency` 会根据集群配置自动调整，无需手动设置。如果通过 Grafana 中的 **TiKV-Details** > **Backup & Import** > **Import RPC count** 监控指标发现 download 文件数量长时间接近于 0，而 ingest 文件数量一直处于上限时，说明 ingest 文件任务存在堆积，并且作业队列已达到最大长度。此时，可以采取以下措施来缓解积压问题：

    - 设置 `--ratelimit` 参数来限制下载速度，以确保 ingest 文件任务有足够的资源。例如，当任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽大于 `x/2 MiB/s`，可以设置参数 `--ratelimit x/2`。如果任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽小于或等于 `x/2 MiB/s`，可以不设置参数 `--ratelimit`。
    - 调高 `--tikv-max-restore-concurrency` 来增加作业队列的最大长度。
```

Comment (qiancai) [inline br/br-snapshot-guide.md:225]:
```suggestion
    通常情况下，`--tikv-max-restore-concurrency` 会根据集群配置自动调整，无需手动设置。如果通过 Grafana 中的 **TiKV-Details** > **Backup & Import** > **Import RPC count** 监控指标发现 download 文件数量长时间接近于 0，而 ingest 文件数量一直处于上限时，说明 ingest 文件任务存在堆积，并且作业队列已达到最大长度。此时，可以采取以下措施来缓解任务堆积问题：

    - 设置 `--ratelimit` 参数来限制下载速度，以确保 ingest 文件任务有足够的资源。例如，当任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽大于 `x/2 MiB/s`，可以设置参数 `--ratelimit x/2`。如果任意 TiKV 节点的硬盘吞吐量为 `x MiB/s` 且下载备份文件的网络带宽小于或等于 `x/2 MiB/s`，可以不设置参数 `--ratelimit`。
    - 调高 `--tikv-max-restore-concurrency` 来增加作业队列的最大长度，缓解任务堆积问题。
```

File: br/br-snapshot-guide.md [hunk 9]
```diff
@@ -217,6 +217,13 @@ TiDB 备份功能对集群性能（事务延迟和 QPS）有一定的影响，
     --log-file restorefull.log
     ```
 
+- 从 v8.0.0 起，`br` 命令行工具新增 `--tikv-max-restore-concurrency` 参数，用于控制每个 TiKV 节点的最大 download 和 ingest 文件数量。此外，通过调整此参数，可以控制作业队列的最大长度（作业队列的最大长度 = 32 * TiKV 节点数量 * `--tikv-max-restore-concurrency`），进而控制 BR 节点的内存消耗。
```
Comment (qiancai) [inline br/br-snapshot-guide.md:220]:
```suggestion
- 从 v8.0.0 起，`br` 命令行工具新增 `--tikv-max-restore-concurrency` 参数，用于控制每个 TiKV 节点的最大 download 和 ingest 文件数量。此外，通过调整此参数，可以控制作业队列的最大长度（作业队列的最大长度 = 32 \* TiKV 节点数量 \* `--tikv-max-restore-concurrency`），进而控制 BR 节点的内存消耗。
```

**Pull request merged**